utils/regr/moda.h

```diff
@@ -67,34 +70,77 @@ struct hasher
   utils::Hasher fHasher;
 };
 
+// A special hasher for double that may only have 10 bytes
 template <>
 struct hasher<long double>
 {
+  hasher<long double>(datatypes::Charset cs)
+  {}
   inline size_t operator()(long double val) const
   {
-    if (sizeof(long double) == 8)  // Probably just MSC, but you never know.
-    {
-      return fHasher((char*)&val, sizeof(long double));
-    }
-    else
-    {
-      // For Linux x86_64, long double is stored in 128 bits, but only 80 are significant
-      return fHasher((char*)&val, 10);
-    }
+#ifdef MASK_LONGDOUBLE
+    // For Linux x86_64, long double is stored in 128 bits, but only 80 are significant
+    return fHasher((char*)&val, 10);
+#else
+    return fHasher((char*)&val, sizeof(long double));
+#endif
   }
 
  private:
   utils::Hasher fHasher;
 };
 
+// A collation aware hasher for strings
+template<>
+struct hasher<string>
+{
+  hasher<string>(datatypes::Charset cs) : fHasher(cs){}
+  inline size_t operator()(string val) const
+  {
+    return fHasher(val.c_str(), val.size());
+  }
+
+private:
+  hasher<string>() : fHasher(datatypes::Charset(8)) {} // Private makes disabled
+  datatypes::CollationAwareHasher fHasher; 
+};
+
+template<class T>
+struct comparator
+{
+  comparator(datatypes::Charset cs) {}
+
+  bool operator()(const T& lhs, const T& rhs) const
+  {
+    return lhs == rhs;
+  }
+};
+// A collation aware string comparator
+template <>
+struct comparator<std::string>
```

Comment:
We are getting into the same trouble here producing another collation aware comparator. The only thing that stops me from suggesting ConstString here is the fact that ctor ConstString costs something. But I would rather use MariaDBHasher/MariaDBComparator then produce another Hasher/Comparator pair for a hashmap.

Reply:
I'm not sure why this is a problem here. I need a template based hasher and comparator for the map. Simple specialization for string type using the charset.eq() operator and already existing CollationAwareHasher seems simple and extremely lightweight.